crates/symbolicator/src/cache.rs

```diff
@@ -34,9 +36,11 @@ pub enum CacheStatus {
     /// A cache item that represents the absence of something. E.g. we encountered a 404 or a client
     /// error while trying to download a file, and cached that fact. Represented by an empty file.
     Negative,
-    /// We are unable to create or use the cache item. E.g. we failed to create a symcache, or
-    /// encountered an error while downloading a file. See docs for [`MALFORMED_MARKER`].
+    /// We are unable to create or use the cache item. E.g. we failed to create a symcache. See docs
+    /// for [`MALFORMED_MARKER`].
     Malformed,
+    /// couldn't download the thing
+    DownloadError,
```

Comment:
I think this should be `Error` and not `DownloadError`. This cache status is generic and also for the derived caches like cficache.  A download error for a derived cache would probably be something like error with a message of `failed to fetch object: <orginal message>`.  But it could also be that the conversion fails due to a bug in symbolic.  In that case it'd also be an `Error` but no a download error.

Reply:
The idea of having this be a catch-all that includes unexpected errors such as bugs in symbolic gives me pause, as it would mean that they'd use the same expiration strategy as download errors, i.e. allowing the download to be retried after an hour. I figure with symbolic errors it'd make more sense to expire them in the same way we expire `Malformed` items, namely after symbolicator restarts. If it's possible to switch to a new version of symbolic without restarting symbolicator then this probably is a moot point.

Exploring the idea of just stuffing all of the errors into `DownloadError`/`Error`, it's certainly doable to differentiate between download errors and other non-download related errors. This would probably mean adding in another error variant or nesting a second enum in `Error` in the same way #502 is extending `Malformed`. This would most likely make the changes in this a loose superset of those in #502, making it harder to justify continuing with this solution I think.

Reply:
After double checking the code paths a bit, I'm a little confused. If I understand correctly, the concern is that `CacheStatus` is used for the data cache, the cfi cache, and the sym cache but the derived (cfi+sym) caches may not have a use for a `DownloadError` status. To ensure that all variants are useful to the caches using them, `DownloadError` should be extended to `Error` so the derived caches can use it to capture conversion failures?

If I understand the `compute` implementations for the caches, derived caches are already using `CacheStatus::Malformed` if conversion fails. Could I be misunderstanding what you're asking for or the code itself? Are you interested in being able to differentiate between conversion errors and original objects that are malformed? If that happens to be the case I'm unsure if that falls under the scope of this change as it's primarily concerned with download errors,  although it would be heavily related.

If the concern is that `DownloadError` is being used for conversion errors due to a lack of visibility on the root cause of an error related to cache computation, I do believe that all conversion issues are emitted as `CacheStatus::Malformed` entries. Some download errors are being turned negative cache entries, but this happens when we're receiving an error message that suggests that the file is missing in some way.

I do think that I might need to double check some details to make sure the right data is written to the appropriate caches (if the fetch errors are different between the data cache and the derived cache, what do we do? do we want to propagate the original download error message into the derived caches?) but I'm unsure if that's what you're referring to here?